### ISSUE TITLE
New file format with named (or numbered) features

### ISSUE THREAD

**Issue by jamespharaoh** [zbackup/zbackup#141]
In order to start evolving the zbackup file format, I think we need to bump the file format version number. Rather than doing this in a rigid way, I propose that we add a table of features which can be enabled per repository, or even per file.

Presumably the next version number would be 2 (is anyone using a version number other than 1 right now?), then the features can either be named or assigned numbers centrally. I kind of prefer the names since this would allow people to create their own features with a namespace without some kind of central administration. Perhaps these could be enumerated in a core repository file and then referenced by number in individual files, so that we aren't repeating a long string many times.

I could then implement the two features I want which require a new disk format - #131 and #140, for example using the feature names x-wbs-chunk-range and x-wbs-multipart-bundles, and if these were accepted by the zbackup core community the names could be standardised to something similar, in a similar way that HTTP and CSS features are added by browsers with custom prefixes before standardisation.

**Comment by mknjc** [comment]
Please don't add feature flags. Its a backup program and not a website where webbrowser can ignore css tags they don't know. Where should zbackup know if its able to restore the backup correctly when it sees feature flags it didn't support?
My recommendation for out of tree features:
If its a backward compatible feature like additional indices add them in own folders in the backup dir and generate the data from the main data if your index is missing.
If its not backward compatible, its not zbackup. Try find Am1GO and ask him to add the feature into zbackup, fork zbackup or just don't name your software zbackup.

**Comment by jamespharaoh** [comment]
The main feature I want to add is ranged chunk restorage - ie restore chunk xxx from byte y to z.

This is probably directly useful in the core zbackup itself, since it can't currently deduplicate overlapping chunks. If zbackup now finds an overlapping chunk it can only reference one of them, so surely this could not only be implemented for restore but also made a part of backup, in a both useful and testable way.

The named feature idea was to allow extra extensions to be used in a dynamic way, where the incompatibility could be detected, and allow extra features to be used in a transparent way. If there's resistance against that, it's no problem, but it seems like a fairly simple way to "detect" incompatibilities while allowing extensions. This would be combined with a version number change, and a strict requirement for flags to be marked as "required", so it would be trivial to know if a feature was not supported, assuming some kind of registry or global naming convention.

In any case, my software is called rzbackup, and I've already asked permission from the original author of zbackup to use that name.

I'm doing my best to make my file format 100% compatible with zbackup, but the advantages of ranged chunk restore are potentially /enormous/, based on a lot of tests I've done, and I'm keen for my repositories to still be compatible. I have some more outrageous ideas also, and would rather any repository that uses them to be easily identifiable by a "compliant" implementation rather than confusing users.

**Comment by mknjc** [comment]
Yes I do understand why you like to see feature flags, but think the user way: Think about a zip file with a special flag set by your software but using cool delta features, it looks like a zip file but no unzip tool can extract files without errors, there are only two possible ways to handle unknown flags: ignore them and try to unzip, which might fail or produces wrong output, maybe even without a message. -> Very bad, user has broken data and didn't know about.
Or fail on unknown flags which causes user hadaches because unzip ./bla.zip responds with: unknown flag flag-cust-cool-delta and didn't extract anything. -> Also bad because user don't know how to extract the files.
The better way would be: name this file bla.cust. The user checks what software can open .cust files and is happy.

I only see two scenarios where feature flags are useful:
1. With a direct connection between a server and a client where they could deal a featureset. See TLS with Cipher selection.
2. With features which could be ignored and only misses speed/visual features like a additional index or the already mentions css attributes. Its not so much hassle if your backup software can't display the xattr of a file because it didn't have the feature. But if zbackup is unable to open a zbackup file users don't know what to do.

>In any case, my software is called rzbackup, and I've already asked permission from the original author of zbackup to use that name.

Sorry I didn't want to insult you. I meant you shouldn't use the name zbackup for your files if they aren't compatible with zbackup not you aren't allowed to use the name.

**Comment by jamespharaoh** [comment]
Yeah, I understand the issues around feature flags, but I thought it would be a good way to try out new features which, from various testing I've done, could make zbackup much more efficient. I am trying to work out how to do this with the least impact on zbackup users, and imagined that a "required" but "unrecognised" feature flag would be a good way to use it.

I agree, making it clear that it it NOT compatible with zbackup is essential and would only enable these features behind flags that make this clear, and am also considering the important process of testing out new features in the wild. Making sure users are aware of the incompatible features, giving them a relatively useful feature if they try to restore from a repository with them, and allowing "optional" features to be added in a way that can be recognised by tools which understand them but ignored by those which will work without them, is my goal.

ZBackup already checks the version number of files it extracts, so it would immediately tell a user if they were using an incompatible repository. Having a single version number increase with some logic to add extra features with optional compatibility would potentially avoid the necessity to increase it again while still allowing optional features to be used by an implementation which doesn't support them with no data loss, presumably a loss of performance etc, and allow new features to be tested widely in a safe way.
